### PR TITLE
BACKLOG-35131 Azure HDInsights Shim

### PR DESCRIPTION
**Pull request by renjith-nem**

### CODE REVIEW

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 17m 23s
Build command:
```console
mvn clean verify -B -e -Daudit -amd
```
---
:ok_hand: All tests passed!
 
Tests run: **930**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/big-data-hadoop-shims/job/PR-1210/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 17m 6s
Build command:
```console
mvn clean verify -B -e -Daudit -amd
```
---
:ok_hand: All tests passed!
 
Tests run: **930**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/big-data-hadoop-shims/job/PR-1210/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 16m 47s
Build command:
```console
mvn clean verify -B -e -Daudit -amd
```
---
:ok_hand: All tests passed!
 
Tests run: **930**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/big-data-hadoop-shims/job/PR-1210/3/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 17m 13s
Build command:
```console
mvn clean verify -B -e -Daudit -amd
```
---
:ok_hand: All tests passed!
 
Tests run: **930**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/big-data-hadoop-shims/job/PR-1210/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 16m 11s
Build command:
```console
mvn clean verify -B -e -Daudit -amd
```
---
:ok_hand: All tests passed!
 
Tests run: **930**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/big-data-hadoop-shims/job/PR-1210/5/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 17m 20s
Build command:
```console
mvn clean verify -B -e -Daudit -amd
```
---
:ok_hand: All tests passed!
 
Tests run: **930**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/big-data-hadoop-shims/job/PR-1210/6/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 16m 40s
Build command:
```console
mvn clean verify -B -e -Daudit -amd
```
---
:ok_hand: All tests passed!
 
Tests run: **930**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/big-data-hadoop-shims/job/PR-1210/7/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: shims/hdi40/driver/src/main/java/org/pentaho/hadoop/shim/hdi/format/HDIFormatShim.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import org.pentaho.hadoop.shim.api.format.IPentahoParquetOutputFormat;
+import org.pentaho.hadoop.shim.common.CommonFormatShim;
+import org.pentaho.hadoop.shim.common.format.avro.PentahoAvroInputFormat;
+import org.pentaho.hadoop.shim.common.format.avro.PentahoAvroOutputFormat;
+import org.pentaho.hadoop.shim.hdi.format.orc.HDIOrcInputFormat;
+import org.pentaho.hadoop.shim.hdi.format.orc.HDIOrcOutputFormat;
+import org.pentaho.hadoop.shim.hdi.format.parquet.HDIApacheInputFormat;
+import org.pentaho.hadoop.shim.hdi.format.parquet.HDIApacheOutputFormat;
+
+public class HDIFormatShim extends CommonFormatShim {
```
Comment (renjith-nem) [inline shims/hdi40/driver/src/main/java/org/pentaho/hadoop/shim/hdi/format/HDIFormatShim.java:41]: we've had to introduce a separate format + implementation for Orc and Parquet as opposed to the common format class for HDInsights, as we have additional steps to pass in the correct file system for the reader/writer and also for the makequalified funcationality as the HDInsights cluster storage pattern is different as opposed to other shims.

Ex: 

> <property>
      <name>fs.defaultFS</name>
      <value>wasb://wasb-secure-cluster-2021-05-17t09-37-48-203z@wasbclustersehdistorage.blob.core.windows.net</value>
      <final>true</final>
    </property>

Around 80% of code is the same as that in CommonShim classes however, we cannot extend the classes unless we make few of the fields protected under common-fragment repo to allow to be used under shim project and scope of change would be larger and extent to all other shims. With this, the change is limited to only HDInsights shim with code duplication being the only drawback.

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 13m 51s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
shims,shims/hdi40,shims/hdi40/driver,shims/hdi40/kar,shims/hdi40/pmr
```
---
:ok_hand: All tests passed!
 
Tests run: **35**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/big-data-hadoop-shims/job/PR-1210/8/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: shims/hdi40/driver/src/main/java/org/pentaho/hadoop/shim/hdi/format/orc/HDIOrcRecordReader.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    //Create a map of input fields to Orc Column numbers
+    schemaToOrcSubcripts = new HashMap<>();
+    for ( IOrcInputField inputField : dialogInputFields ) {
+      if ( inputField != null ) {
+        Integer colNumber = orcColumnNumberMap.get( inputField.getFormatFieldName() );
+        if ( colNumber == null ) {
+          throw new IllegalArgumentException(
+                  "Column " + inputField.getFormatFieldName()
```
Comment (ddiroma) [inline shims/hdi40/driver/src/main/java/org/pentaho/hadoop/shim/hdi/format/orc/HDIOrcRecordReader.java:95]: @renjith-nem Should the strings be localized?

File: shims/hdi40/driver/src/main/java/org/pentaho/hadoop/shim/hdi/format/parquet/HDIApacheOutputFormat.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+      TaskAttemptID taskAttemptID = new TaskAttemptID( "qq", 111, TaskType.MAP, 11, 11 );
+      TaskAttemptContextImpl task = new TaskAttemptContextImpl( job.getConfiguration(), taskAttemptID );
+      try {
+
+        ParquetRecordWriter<RowMetaAndData> recordWriter =
+                (ParquetRecordWriter<RowMetaAndData>) nativeParquetOutputFormat.getRecordWriter( task );
+        return new PentahoParquetRecordWriter( recordWriter, task );
+      } catch ( IOException e ) {
+        throw new IllegalStateException( "Some error accessing parquet files", e );
```
Comment (ddiroma) [inline shims/hdi40/driver/src/main/java/org/pentaho/hadoop/shim/hdi/format/parquet/HDIApacheOutputFormat.java:192]: @renjith-nem Should these strings be localized?

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 16m 56s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
common-fragment-V1,shims,shims/hdi40,shims/hdi40/driver,shims/hdi40/kar,shims/hdi40/pmr
```
---
:ok_hand: All tests passed!
 
Tests run: **814**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/big-data-hadoop-shims/job/PR-1210/9/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 15m 41s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
common-fragment-V1,shims,shims/hdi40,shims/hdi40/driver,shims/hdi40/kar,shims/hdi40/pmr
```
---
:ok_hand: All tests passed!
 
Tests run: **814**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/big-data-hadoop-shims/job/PR-1210/10/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 16m 7s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
common-fragment-V1,shims,shims/hdi40,shims/hdi40/driver,shims/hdi40/kar,shims/hdi40/pmr
```
---
:ok_hand: All tests passed!
 
Tests run: **814**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/big-data-hadoop-shims/job/PR-1210/11/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 16m 56s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
common-fragment-V1,shims,shims/hdi40,shims/hdi40/driver,shims/hdi40/kar,shims/hdi40/pmr
```
---
:ok_hand: All tests passed!
 
Tests run: **814**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/big-data-hadoop-shims/job/PR-1210/12/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Pull request merged**